### ISSUE TITLE
[Question] What is the correct code to skip `Para` and `Plain` for supporting Codebraid Preview with `sourcepos`?

### ISSUE THREAD

**Issue by gusbemacbe** [gpoore/codebraid-preview-vscode#21]
Good morning!

I wrote two Pandoc Lua filters – admonition and collapsible section. They work only with the `markdown_github` variant. Not with the `commonmark_x` variant because of `data-pos` and sourcepos, and `Para` or `Plain` are like `Div`. I acknowledged your note on filters. 

I was studying @wlupton's old and new old `list-table` codes to understand better how to add support for Codebraid Preview and for sourcepos. They almost worked, but I failed when I changed to a content filter. 

Here is the Markdown admonition syntax code:

```md
::: note
This is a _note_ with a `code`

It has _multiple_ paragraphs for `code`.
:::
```

It will become into HTML:

```html
<div class="admonition note">
  <p class="admonition-title">This is a <i>note</i> with a <code>code</code></p>

  <p>It has <i>multiple</i> paragraphs for <code>code</code>
</div>
```

I wrote this Lua filter code:

```lua
local function admonition(el)
    if el.classes[1] == "note" or el.classes[1] == "bug" or el.classes[1] == "error" or el.classes[1] == "tip" or
        el.classes[1] == "warning" then
        local title = el.content[1].content
        local pandoc_title = pandoc.Pandoc({pandoc.Para(title)}, pandoc.Meta({}))
        local html_title = pandoc.write(pandoc_title, "html")
        html_title = html_title:gsub("<p>", ""):gsub("</p>", "")
        table.remove(el.content, 1)
        return {pandoc.RawBlock("html",
            '<div class="admonition ' .. el.classes[1] .. '"><p class="admonition-title">' .. html_title .. '</p>'),
                pandoc.Div(el.content), pandoc.RawBlock("html", "</div>")}
    else
        return el
    end
end

return {{Div = admonition}}
```

It failed because of `Para`. I switched from ` local title = el.content[1].content` to `local summary = el.content[1].classes[1]`, and it almost worked on Codebraid Preview with `commonmark_x`, but the problem is that it will display `note` instead of `This is a note with a code`.

I was looking at @wlupton's codes about `block_skip_data_pos`, and I adapted my code, but it didn't work yet.

Here is the Markdown collapsible section syntax code, similar to that below:

```md
::: collapse
Attention to the `code`

Here is some *hidden* **content**
:::
```

And to HTML:

```html
<details>
   <summary>Attention to the <code>code</code></summary>
   Here is some <i>hidden</i> <b>content</b>
</details
```

Here is its Lua code:

```lua
local function collapsible_section(el)
    if el.classes[1] == "collapse" then
        local summary = el.content[1].content
        local pandoc_summary = pandoc.Pandoc({pandoc.Para(summary)}, pandoc.Meta({}))
       local html_summary = pandoc.write(pandoc_summary, "html")
        html_summary = html_summary:gsub("<p>", ""):gsub("</p>", "")
        table.remove(el.content, 1)
        local details = pandoc.Div(el.content)
        return {pandoc.RawBlock("html", "<details><summary>" .. html_summary .. "</summary>"), details,
                pandoc.RawBlock("html", "</details>")}
    else
        return el
    end
end

return {{ Div = collapsible_section }}
```

**Comment by wlupton** [comment]
I recommend using my https://github.com/pandoc-ext/logging to show the AST (I also recommend using _very_ simple examples, so as to keep the AST simple). For example, with this `note.md`:
```
::: note
first

second
:::
```

and this variant of your `admonition.lua`:
```
local logging = require 'logging'
local temp = logging.temp

local function admonition(el)
    if el.classes[1] == "note" or el.classes[1] == "bug" or el.classes[1] == "error" or el.classes[1] == "tip" or
        el.classes[1] == "warning" then
        temp('el', el)
        local title = el.content[1].content
        local pandoc_title = pandoc.Pandoc({pandoc.Para(title)}, pandoc.Meta({}))
        local html_title = pandoc.write(pandoc_title, "html")
        html_title = html_title:gsub("<p>", ""):gsub("</p>", "")
        table.remove(el.content, 1)
        return {pandoc.RawBlock("html",
            '<div class="admonition ' .. el.classes[1] .. '"><p class="admonition-title">' .. html_title .. '</p>'),
                pandoc.Div(el.content), pandoc.RawBlock("html", "</div>")}
    else
        return el
    end
end

return {{Div = admonition}}
```

you get this output:
```
% pandoc -f commonmark_x+sourcepos note.md -L admonition.lua
(#) el Div {
  attr: Attr {
    attributes: AttributeList {
      data-pos: "note.md@1:1-6:1"
    }
    classes: List[1] {
      [1] "note"
    }
    identifier: ""
  }
  content: Blocks[2] {
    [1] Div {
      attr: Attr {
        attributes: AttributeList {
          data-pos: "note.md@2:1-3:1"
        }
        classes: List {}
        identifier: ""
      }
      content: Blocks[1] {
        [1] Para {
          content: Inlines[1] {
            [1] Span {
              attr: Attr {
                attributes: AttributeList {
                  data-pos: "note.md@2:1-2:6"
                }
                classes: List {}
                identifier: ""
              }
              content: Inlines[1] {
                [1] Str "first"
              }
            }
          }
        }
      }
    }
    [2] Div {
      attr: Attr {
        attributes: AttributeList {
          data-pos: "note.md@4:1-5:1"
        }
        classes: List {}
        identifier: ""
      }
      content: Blocks[1] {
        [1] Para {
          content: Inlines[1] {
            [1] Span {
              attr: Attr {
                attributes: AttributeList {
                  data-pos: "note.md@4:1-4:7"
                }
                classes: List {}
                identifier: ""
              }
              content: Inlines[1] {
                [1] Str "second"
              }
            }
          }
        }
      }
    }
  }
}
Error running filter admonition.lua:
Inline, list of Inlines, or string expected, got Blocks
	while retrieving function argument content
	while retrieving arguments for function Para
stack traceback:
	admonition.lua:9: in function <admonition.lua:4>
```

I hope that it's clear that you need to do something like this:
```
local content = blocks_skip_data_pos(el.content)
```

and then, in the rest of the filter, where you previously had `el.content` use `content` instead.

When I tried this I got this output:
```
<div class="admonition note"><p class="admonition-title"><span data-pos="note.md@2:1-2:6">first</span></p>
<div>
<p><span data-pos="note.md@4:1-4:7">second</span></p>
</div>
</div>
```

**Comment by gpoore** [comment]
I think @wlupton has covered the basics of what is needed for `sourcepos` compatibility.

In the next release, I will be adding an option to disable `sourcepos` for `commonmark_x` and use the same scroll sync approach that is used for non-CommonMark formats. This will sacrifice some scroll sync accuracy, but it will provide a way to quickly get filters working with `commonmark_x` without needing to worry about the `data-pos` AST elements.

**Comment by gusbemacbe** [comment]
@wlupton and @gpoore 

I tested what you suggested, and it worked very well, thank you.

**Comment by gpoore** [comment]
I've just released v0.16. Among other things, this adds a new setting `codebraid.preview.pandoc.preferPandocSourcepos`. Disabling this will often make `commonmark`, `commonmark_x`, and `gfm` compatible with filters without requiring any modifications. Pandoc's `sourcepos` usually gives better scroll sync, but also inserts lots of `data-pos` AST elements. Codebraid Preview's `sourcepos` may give less accurate scroll sync, but it introduces fewer AST modifications (and very few block-level `data-pos` elements) so it is usually more compatible with filters.
